contracts/cw721-base/src/contract.rs

```diff
@@ -14,442 +12,609 @@ use cw721::{
 
 use crate::error::ContractError;
 use crate::msg::{ExecuteMsg, InstantiateMsg, MintMsg, MinterResponse, QueryMsg};
-use crate::state::{
-    increment_tokens, num_tokens, tokens, Approval, TokenInfo, CONTRACT_INFO, MINTER, OPERATORS,
-};
-use cw_storage_plus::Bound;
+use crate::state::{token_owner_idx, Approval, TokenIndexes, TokenInfo};
+use cw_storage_plus::{Bound, IndexedMap, Item, Map, MultiIndex};
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::marker::PhantomData;
 
 // version info for migration info
 const CONTRACT_NAME: &str = "crates.io:cw721-base";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
-#[cfg_attr(not(feature = "library"), entry_point)]
-pub fn instantiate(
-    deps: DepsMut,
-    _env: Env,
-    _info: MessageInfo,
-    msg: InstantiateMsg,
-) -> StdResult<Response> {
-    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
-
-    let info = ContractInfoResponse {
-        name: msg.name,
-        symbol: msg.symbol,
-    };
-    CONTRACT_INFO.save(deps.storage, &info)?;
-    let minter = deps.api.addr_validate(&msg.minter)?;
-    MINTER.save(deps.storage, &minter)?;
-    Ok(Response::default())
+const DEFAULT_LIMIT: u32 = 10;
+const MAX_LIMIT: u32 = 30;
+
+pub struct Cw721Contract<'a, T, C>
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    pub contract_info: Item<'a, ContractInfoResponse>,
+    pub minter: Item<'a, Addr>,
+    pub token_count: Item<'a, u64>,
+    /// Stored as (granter, operator) giving operator full control over granter's account
+    pub operators: Map<'a, (&'a Addr, &'a Addr), Expiration>,
+    pub tokens: IndexedMap<'a, &'a str, TokenInfo<T>, TokenIndexes<'a, T>>,
```

Comment:
I like the idea of having state as part of implementation structure as a part of contract type itself - I didn't put it in the original proposal as I wanted to make it minimal, but if it is achievable then I think it is a good idea.

Reply:
It's not just achievable, it is essential in this use-case to allow the state itself to be generic over T and ensure all state access uses the same T